File: libs/shared/src/ai_circus_shared/form_validation.py
"""Generic, domain-agnostic field/submission validation for `assisted_form` scenarios.

Every rule here is one of a small, reusable set of primitives (email/phone/pattern/
min_length) — a scenario supplies the domain-specific detail (e.g. a national ID
number's format) as plain data (`pattern`) in its `scenario.yaml`, never as new code
here. `ui-react`'s `formValidation.ts` mirrors these same rules client-side for
instant feedback; this module is the final authority at submission time (see
`form-agent`'s `POST /submissions/{slug}`).
"""

from __future__ import annotations

import re

from ai_circus_shared.scenario_schema import FormConfig, FormFieldSpec

_EMAIL_RE = re.compile(r"^[^@\s]+@[^@\s]+\.[^@\s]+$")
_PHONE_RE = re.compile(r"^\+?[0-9 ()-]{7,20}$")


def validate_field(spec: FormFieldSpec, value: str) -> str | None:
    """Validate one non-empty field value against its `validation` rule.

    Returns an error message, or `None` if valid. Required-ness is `validate_submission`'s
    job, not this function's — an empty value is always valid here.
    """
    if not value:
        return None
    if spec.validation == "email" and not _EMAIL_RE.match(value):
        return "Must be a valid email address."
    if spec.validation == "phone" and not _PHONE_RE.match(value):
        return "Must be a valid phone number."
    if spec.validation == "pattern":
        assert spec.pattern is not None  # enforced by FormFieldSpec's own validator
        if not re.match(spec.pattern, value):
            return spec.helper_text or "Does not match the required format."
    if spec.validation == "min_length":
        assert spec.min_length is not None  # enforced by FormFieldSpec's own validator
        if len(value) < spec.min_length:
            return f"Must be at least {spec.min_length} characters."
    return None


def _is_required(spec: FormFieldSpec, fields: dict[str, str]) -> bool:
    if spec.required:
        return True
    if spec.required_if is not None:
        return fields.get(spec.required_if.field) in spec.required_if.in_values
    return False


def validate_submission(form: FormConfig, fields: dict[str, str]) -> dict[str, str]:
    """Validate a full submission against `form`'s field specs.

    Returns `{field_id: error_message}` for every missing-required or invalid field —
    an empty dict means the submission is ready to persist.
    """
    errors: dict[str, str] = {}
    for spec in form.fields:
        value = fields.get(spec.id, "").strip()
        if not value:
            if _is_required(spec, fields):
                errors[spec.id] = "This field is required."
            continue
        error = validate_field(spec, value)
        if error is not None:
            errors[spec.id] = error
    return errors

File: ui-react/src/formValidation.ts
/**
 * Client-side mirror of ai_circus_shared.form_validation — instant per-field
 * feedback and the Submit button's enabled state. The backend (form-agent's
 * POST /submissions/{slug}) is the final authority at submit time; this exists only
 * so the user doesn't have to round-trip to find out a field is missing/invalid.
 */
import type { FormConfig, FormFieldSpec } from "./apiClient";

const EMAIL_RE = /^[^@\s]+@[^@\s]+\.[^@\s]+$/;
const PHONE_RE = /^\+?[0-9 ()-]{7,20}$/;

/** Validate one non-empty field value against its `validation` rule. Required-ness
 * is `validateForm`'s job, not this function's — an empty value is always valid here.
 */
export function validateField(spec: FormFieldSpec, value: string): string | null {
  if (!value) return null;
  if (spec.validation === "email" && !EMAIL_RE.test(value)) return "Must be a valid email address.";
  if (spec.validation === "phone" && !PHONE_RE.test(value)) return "Must be a valid phone number.";
  if (spec.validation === "pattern" && spec.pattern && !new RegExp(spec.pattern).test(value)) {
    return spec.helper_text ?? "Does not match the required format.";
  }
  if (spec.validation === "min_length" && spec.min_length != null && value.length < spec.min_length) {
    return `Must be at least ${spec.min_length} characters.`;
  }
  return null;
}

function isRequired(spec: FormFieldSpec, values: Record<string, string>): boolean {
  if (spec.required) return true;
  if (spec.required_if) return spec.required_if.in_values.includes(values[spec.required_if.field] ?? "");
  return false;
}

/** Validate every field in `form` against `values`. Returns `{field_id: message}` for
 * every missing-required or invalid field — an empty object means the form is ready
 * to submit.
 */
export function validateForm(form: FormConfig, values: Record<string, string>): Record<string, string> {
  const errors: Record<string, string> = {};
  for (const spec of form.fields) {
    const value = (values[spec.id] ?? "").trim();
    if (!value) {
      if (isRequired(spec, values)) errors[spec.id] = "This field is required.";
      continue;
    }
    const error = validateField(spec, value);
    if (error) errors[spec.id] = error;
  }
  return errors;
}
